File: mpp/common/src/commonMain/kotlin/Repository.kt
package mpp

import io.ktor.client.HttpClient
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.response.HttpResponse
import io.ktor.client.response.readText
import io.ktor.http.isSuccess
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.list

@Serializable
data class Pet(val externalId: Int?, val title: String?)

@Serializable
data class AdoptersLastUpdated(val adoptersLastUpdated: Long?)

@Serializable
data class PetName(val name: String?, val lastUpdateTime: Long?)

suspend fun getPets(client: HttpClient, baseUrl: String, adopter: String): List<Pet>? {
    val response = client.get<HttpResponse>("${baseUrl}api/petsByAdopter") {
        url {
            parameters.append(name = "adopter", value = adopter)
        }
    }
    if (!(response.status.isSuccess())) {
        return null
    }
    val responseText = response.readText()
    return Json.nonstrict.parse(deserializer = Pet.serializer().list, string = responseText)
}

suspend fun adoptPet(client: HttpClient, baseUrl: String, privateKey: String, petId: Int): Boolean {
    val response = client.post<HttpResponse>("${baseUrl}api/adopt") {
        url {
            parameters.append(name = "privateKey", value = privateKey)
            parameters.append(name = "petId", value = petId.toString())
        }
    }
    return response.status.isSuccess()
}

suspend fun assignNameToPet(client: HttpClient, baseUrl: String, privateKey: String, petId: Int, name: String): Boolean {
    val response = client.post<HttpResponse>("${baseUrl}api/assignName") {
        url {
            parameters.append(name = "privateKey", value = privateKey)
            parameters.append(name = "petId", value = petId.toString())
            parameters.append(name = "name", value = name)
        }
    }
    return response.status.isSuccess()
}

suspend fun getLastKnownAdoptersUpdate(client: HttpClient, baseUrl: String): AdoptersLastUpdated? {
    val response = client.get<HttpResponse>("${baseUrl}api/lastKnownAdoptersUpdate")
    if (!(response.status.isSuccess())) {
        return null
    }
    val responseText = response.readText()
    return Json.nonstrict.parse(deserializer = AdoptersLastUpdated.serializer(), string = responseText)
}

suspend fun getPetName(client: HttpClient, baseUrl: String, petId: Int): PetName? {
    val response = client.get<HttpResponse>("${baseUrl}petName/${petId}")
    if (!(response.status.isSuccess())) {
        return null
    }
    val responseText = response.readText()
    return Json.nonstrict.parse(deserializer = PetName.serializer(), string = responseText)
}

File: cli/src/repository.rs
extern crate reqwest;
extern crate serde;

use serde::{Serialize, Deserialize};

use reqwest::Client;
use reqwest::Url;

use std::error;

#[allow(non_snake_case)]
#[derive(Deserialize, Serialize, Debug)]
pub struct Pet {
    pub externalId: i64,
    pub title: Option<String>,
}

#[allow(non_snake_case)]
#[derive(Deserialize, Serialize, Debug)]
pub struct AdoptersLastUpdated {
    pub adoptersLastUpdated: Option<i64>,
}

#[allow(non_snake_case)]
#[derive(Deserialize, Serialize, Debug)]
pub struct PetName {
    pub name: Option<String>,
    pub lastUpdateTime: Option<i64>,
}

type Result<T> = std::result::Result<T, Box<error::Error>>;

pub fn get_pets_by_adopter(client: &Client, base_url: &Url, adopter: &str) -> Result<Vec<Pet>> {
    let url = Url::parse_with_params(base_url.join("api/petsByAdopter")?.as_str(),
                                     &[("adopter", adopter)])?;

    let mut response = client
        .get(url.as_str()).send()?;

    if !(response.status().is_success()) {
        return Err(response.text()?.into());
    }

    let result: Vec<Pet> = response.json()?;
    return Ok(result);
}

pub fn adopt_pet(client: &Client, base_url: &Url, private_key: &str, pet_id: i64) -> Result<()> {
    let url = Url::parse_with_params(base_url.join("api/adopt")?.as_str(),
                                     &[("petId", pet_id.to_string()), ("privateKey", private_key.to_string())])?;

    let mut response = client
        .post(url)
        .send()?;

    if !(response.status().is_success()) {
        return Err(response.text()?.into());
    }

    return Ok(());
}

pub fn assign_name_to_pet(client: &Client, base_url: &Url, private_key: &str, pet_id: i64, name: &str) -> Result<()> {
	let url = Url::parse_with_params(base_url.join("api/assignName")?.as_str(),
                                     &[("petId", pet_id.to_string()), ("name", name.to_string()), ("privateKey", private_key.to_string())])?;

    let mut response = client
        .post(url)
        .send()?;

    if !(response.status().is_success()) {
        return Err(response.text()?.into());
    }

    return Ok(());
}

pub fn get_last_known_adopters_update(client: &Client, base_url: &Url) -> Result<AdoptersLastUpdated> {
    let url = Url::parse(base_url.join("api/lastKnownAdoptersUpdate")?.as_str())?;

    let mut response = client
        .get(url.as_str()).send()?;

    if !(response.status().is_success()) {
        return Err(response.text()?.into());
    }

    let result: AdoptersLastUpdated = response.json()?;
    return Ok(result);
}

pub fn get_pet_name(client: &Client, base_url: &Url, pet_id: i64) -> Result<PetName> {
    let url = Url::parse(base_url.join("petName/")?.join(&pet_id.to_string())?.as_str())?;

    let mut response = client
        .get(url.as_str()).send()?;

    if !(response.status().is_success()) {
        return Err(response.text()?.into());
    }

    let result: PetName = response.json()?;
    return Ok(result);
}
